File: src/leetcode/m236/LowestCommonAncestorOfABinaryTree.go
package m236

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

/**
 * LeetCode 236
 */
func lowestCommonAncestor(root, p, q *TreeNode) *TreeNode {
	// Base case: if the current node is null or matches either p or q, return the current node
	if root == nil || root == p || root == q {
		return root
	}

	// Recur for left and right subtrees
	left := lowestCommonAncestor(root.Left, p, q)
	right := lowestCommonAncestor(root.Right, p, q)

	// If both left and right are non-null, current node is the LCA
	if left != nil && right != nil {
		return root
	}

	// Otherwise, return non-null node (either left or right)
	if left != nil {
		return left
	}

	return right
}

File: src/leetcode/m236/LowestCommonAncestorOfABinaryTree.java
/*
 * Copyright (c) 2025 THGEEK.COM. All rights reserved.
 */

package leetcode.m236;

import leetcode.TreeNode;

/**
 * LeetCode 236
 *
 * @author Tao Hong
 * @version 1.0
 * @since 2025/05/06 13:27
 */
public class LowestCommonAncestorOfABinaryTree {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        if (root == null || root == p || root == q) {
            return root;
        }
        var left = lowestCommonAncestor(root.left, p, q);
        var right = lowestCommonAncestor(root.right, p, q);
        if (left != null && right != null) {
            return root;
        }
        if (left != null) {
            return left;
        }
        return right;
    }
}
